---
title: "tidySCE - part of tidytranscriptomics"
output: github_document
always_allow_html: true
---

<!-- badges: start -->
[![Lifecycle:maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing) [![R build status](https://github.com/stemangiola/tidySCE/workflows/R-CMD-check-bioc/badge.svg)](https://github.com/stemangiola/tidySCE/actions)
<!-- badges: end -->

```{r echo=FALSE}
knitr::opts_chunk$set( fig.path = "man/figures/")
```

**Brings SingleCellExperiment to the tidyverse!** 

website: [stemangiola.github.io/tidySCE/](https://stemangiola.github.io/tidySCE/)

Please have a look also to 

- [tidyseurat](https://stemangiola.github.io/tidyseurat/) for tidy manipulation of Seurat objects
- [tidybulk](https://stemangiola.github.io/tidybulk/) for tidy high-level data
analysis and manipulation 
- [nanny](https://github.com/stemangiola/nanny) for tidy high-level data
analysis and manipulation 
- [tidygate](https://github.com/stemangiola/tidygate) for adding custom
gate information to your tibble 
- [tidyHeatmap](https://stemangiola.github.io/tidyHeatmap/) for heatmaps produced with tidy principles

<!---

[![Build Status](https://travis-ci.org/stemangiola/tidySCE.svg?branch=master)](https://travis-ci.org/stemangiola/tidySCE) [![Coverage Status](https://coveralls.io/repos/github/stemangiola/tidySCE/badge.svg?branch=master)](https://coveralls.io/github/stemangiola/tidySCE?branch=master)

-->

```{r, echo=FALSE, include=FALSE, }
library(knitr)
knitr::opts_chunk$set(cache = TRUE, warning = FALSE,
                      message = FALSE, cache.lazy = FALSE)

library(dplyr)
library(tidyr)
library(purrr)
library(magrittr)
library(ggplot2)
library(SingleCellExperiment)
library(SingleCellSignalR)
library(scater)
library(scran)
library(igraph)
library(tidyHeatmap)
library(tidySCE)


my_theme = 	
  list(
    scale_fill_brewer(palette="Set1"),
    scale_color_brewer(palette="Set1") ,
    theme_bw() +
  	theme(
  		panel.border = element_blank(),
  		axis.line = element_line(),
  		panel.grid.major = element_line(size = 0.2),
  		panel.grid.minor = element_line(size = 0.1),
  		text = element_text(size=12),
  		legend.position="bottom",
  		aspect.ratio=1,
  		strip.background = element_blank(),
  		axis.title.x  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
  		axis.title.y  = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
  	)
  )

```

## Functions/utilities available

SingleCellExperiment-compatible Functions | Description
------------ | -------------
`all` | After all `tidySCE` is a SingleCellExperiment object, just better

tidyverse Packages | Description
------------ | -------------
`dplyr` | All `dplyr` APIs like for any tibble
`tidyr` | All `tidyr` APIs like for any tibble
`ggplot2` | `ggplot` like for any tibble
`plotly` | `plot_ly` like for any tibble

Utilities | Description
------------ | -------------
`tidy` | Add `tidySCE` invisible layer over a SingleCellExperiment object
`as_tibble` | Convert cell-wise information to a `tbl_df`
`join_transcripts` | Add transcript-wise information, returns a `tbl_df`

## Installation

From CRAN (in submission)
```{r eval=FALSE}
install.packages("tidySCE")
```

From Github
```{r, eval=FALSE}
devtools::install_github("stemangiola/tidySCE")
```

## Create `tidySCE`, the best of both worlds!
This is a SingleCellExperiment object but it is evaluated as tibble. So it is fully compatible both with SingleCellExperiment and tidyverse APIs. 

```{r}
pbmc_small_tidy = tidySCE::pbmc_small %>% tidy()
```

**It looks like a tibble**

```{r}
pbmc_small_tidy
```

**But it is a SingleCellExperiment object after all**

```{r}
pbmc_small_tidy@assays
```
## Annotation polishing using tidyverse
We may have a column that contains the directory each run was taken from. We may want to extract the run/sample name out of it.

```{r}
pbmc_small_polished =
  pbmc_small_tidy %>%
  extract(file, "sample", "../data/([a-z0-9]+)/outs.+", remove = F) 

pbmc_small_polished %>%
  select(sample, everything())
```

## Preliminary plots
We can treat `pbmc_small_polished` effectively as a normal tibble for plotting. 

Here we plot number of transcripts per cell
```{r plot1, cache=TRUE}

pbmc_small_polished %>%
  tidySCE::ggplot(aes(nFeature_RNA, fill=groups)) + 
  geom_histogram() +
  my_theme

```

Here we plot total transcriptional material per cell
```{r plot2, cache=TRUE}

pbmc_small_polished %>%
  tidySCE::ggplot(aes(groups, nCount_RNA, fill=groups)) + 
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.1) +
  my_theme


```

Here we plot abundance of two transcripts for each group
```{r}
pbmc_small_polished %>% 
  join_transcripts(transcripts = c("HLA-DRA" ,     "LYZ" )) %>%
  ggplot(aes(groups, abundance_counts + 1, fill=groups)) + 
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(aes(size=nCount_RNA), alpha=0.5, width = 0.2) + 
  scale_y_log10() + 
  my_theme

```

## Preprocess the dataset
Also you can treat the object as SingleCellExperiment object and proceed with data processing.

```{r preprocess, cache=TRUE}

variable_gens = 
  pbmc_small_polished %>%
  modelGeneVar() %>%
  getTopHVGs(prop=0.1)

pbmc_small_pca = 
  pbmc_small_polished %>% 
  runPCA(   subset_row = variable_gens  )

pbmc_small_pca
```


If a tool is not included in the tidySCE collection, we can use `as_tibble` to permanently convert `tidySCE` into tibble
```{r pc_plot, cache=TRUE}

pbmc_small_pca %>%
  as_tibble %>%
  select(contains("PC"), everything()) %>%
	  GGally::ggpairs(columns = 1:5, ggplot2::aes(colour=groups))

```

## Identify clusters 
We proceed with cluster identification
```{r cluster, cache=TRUE}

pbmc_small_cluster = pbmc_small_pca


# Assigning to the 'colLabels' of the 'sce'.
colLabels(pbmc_small_cluster) =
  pbmc_small_pca %>%
  buildSNNGraph(use.dimred="PCA") %>%
  igraph::cluster_walktrap() %$%
  membership %>%
  as.factor

pbmc_small_cluster %>% select(label, everything())
```

Now we can interrogate the object as if it was a regular tibble data frame
```{r cluster count, cache=TRUE}

pbmc_small_cluster %>%
  tidySCE::count(groups, label)

```

We can identify cluster markers using SingleCellExperiment
```{r}
# Identify markers
marker_genes = 
  pbmc_small_cluster %>%
  findMarkers(groups=pbmc_small_cluster$label) %>% 
  as.list %>% 
  map(~ .x %>% head(10) %>% rownames) %>% 
  unlist

# Plot heatmap
pbmc_small_cluster %>%
  join_transcripts(transcripts = marker_genes) %>% 
  group_by(label) %>% 
  heatmap(transcript, cell, abundance_counts, .scale = "column")
  
```

## Reduce dimensions
We can calculate the first 3 UMAP dimensions using SingleCellExperiment framework 
```{r umap, cache=TRUE}

pbmc_small_UMAP = 
  pbmc_small_cluster %>%
  runUMAP(ncomponents = 3)

```

and we can plot them using 3D plot using plotly
```{r umap plot, eval=FALSE}

pbmc_small_UMAP %>%
	plot_ly(
		x = ~`UMAP_1`,
		y = ~`UMAP_2`, 
		z = ~`UMAP_3`,
		color = ~ label
	) 

```
![screenshot plotly](man/figures/plotly.png)


## Cell type prediction
We can infer cell type identities using singleR

```{r eval=FALSE}
blueprint = SingleR::BlueprintEncodeData()

cell_type_df = 
  pbmc_small_UMAP@assays@data$logcounts %>%
  Matrix::Matrix(sparse = TRUE) %>%
   SingleR::SingleR(
       ref = blueprint ,
       labels = blueprint$label.main,
       method = "single"
   ) %>%
  as.data.frame() %>%
  as_tibble(rownames="cell") %>% 
  select(cell, first.labels)
```

```{r}
pbmc_small_cell_type =
  pbmc_small_UMAP %>%
  left_join(cell_type_df, by="cell")

pbmc_small_cell_type %>%
  tidySCE::select(cell, first.labels, everything())
```
We can easily summarise the results. For example, see how cell type classification overlaps with cluster classification.
```{r}
pbmc_small_cell_type %>%
  count(label, first.labels)
```

We can easily reshape the data for building information-rish faceted plots

```{r}
pbmc_small_cell_type %>%
  
  # Reshaping
  pivot_longer(
    cols=c(label, first.labels), 
    names_to = "classifier", values_to = "label"
  ) %>%
  
  # Plotting
  ggplot(aes(UMAP1, UMAP2, color=label)) +
  geom_point() +
  facet_wrap(~classifier) +
  my_theme

```

We can easily plot gene correlation per cell category, adding multi-layer annotations

```{r}
pbmc_small_cell_type %>% 
  
  # Add mitochondrial abundance
  mutate(mitochondrial = rnorm(n())) %>%
  
  # Plot correlation
  join_transcripts(transcripts = c("CST3" ,     "LYZ" ), shape = "wide") %>%
  ggplot(aes(CST3 +1, LYZ + 1, color=groups, size=mitochondrial)) +
  geom_point() + 
  facet_wrap(~first.labels, scales = "free") +
  scale_x_log10() +
  scale_y_log10() +
  my_theme
```

##  Nested analyses
A powerful tool we can use with tidySCE is `nest`. We can easily perform independent analyses on subsets of the dataset. First we classify cell types in lymphoid and myeloid; then, nest based on the new classification

```{r}
pbmc_small_nested = 
  pbmc_small_cell_type %>%
  filter(first.labels != "Erythrocytes") %>%
  mutate(cell_class = if_else(`first.labels` %in% c("Macrophages", "Monocytes"), "myeloid", "lmphoid")) %>%
  nest(data = -cell_class)

pbmc_small_nested
```

Now we can independently (i) find variable features, (ii) reduce dimensions, and (iii) cluster. Using both tidyverse and SingleCellExperiment seamlessy

```{r}
pbmc_small_nested_reanalysed = 
  pbmc_small_nested %>%
  mutate(data = map(
    data, ~ {
      
      .x = runPCA(.x, subset_row = variable_gens )
      
      variable_gens = 
        .x %>%
        modelGeneVar() %>%
        getTopHVGs(prop=0.3)
      
      colLabels(.x) =
        .x %>%
        buildSNNGraph(use.dimred="PCA") %>%
        igraph::cluster_walktrap() %$%
        membership %>%
        as.factor
      
      .x %>% runUMAP(ncomponents = 3)
      
    }
  )) 

pbmc_small_nested_reanalysed
```

Now we can unnest and plot the new clasification

```{r}
pbmc_small_nested_reanalysed %>%
  
  # Convert to tibble otherwise SingleCellExperiment drops reduced dimensions when unifying data sets.
  mutate(data = map(data, ~ .x %>% as_tibble)) %>%
  unnest(data) %>%

  # Define unique clusters
  unite("cluster", c(cell_class, label), remove=FALSE) %>%
  
  # Plotting
  ggplot(aes(UMAP1, UMAP2, color=cluster)) +
  geom_point() +
  facet_wrap(~cell_class) +
  my_theme


```

We can perform a large number of functional analyses on data subsets. For example intra-sample cell-cell interactions, for then comparing/testing whether interactions have strenghten/weaken across conditions. In this small dataset we have just two sample (one for condition) but you can imagine how you can use tidyverse to perform t-tests and visualisation.

```{r, eval = FALSE}
pbmc_small_nested_interactions = 
  pbmc_small_nested_reanalysed %>%
  
  # Unnest based on cell category
  unnest(data) %>%
  
  # Create unambiguous clusters
  mutate(integrated_clusters = first.labels %>% as.factor %>% as.integer) %>%

  # Nest based on sample
	tidySCE::nest(data = -sample) %>%
	tidySCE::mutate(interactions = map(data, ~ {
		
		# Produce variables. Yuck!
		cluster = .x@colData$integrated_clusters
		data = data.frame(.x@assays@data %>% as.list %>% .[[1]] %>% as.matrix)

		# Ligand/Receptor analysis using SingleCellSignalR
		data %>%
			cell_signaling(genes=rownames(data),cluster=cluster) %>%
			inter_network(data = data, signal = ., genes = rownames(data), cluster = cluster) %$%
			`individual-networks` %>% 
			map_dfr(~bind_rows(as_tibble(.x))) 
	})) 

pbmc_small_nested_interactions %>%
  select(-data) %>%
  unnest(interactions)
```

If the data set was not so small, and interactions could be identified, you would see something as below.

```{r, echo = FALSE}
tidySCE::pbmc_small_nested_interactions
```
